lib/src/sources/dmi_source.cc

```diff
@@ -0,0 +1,207 @@
+#include <internal/sources/dmi_source.hpp>
+#include <leatherman/util/regex.hpp>
+#include <leatherman/logging/logging.hpp>
+#include <leatherman/file_util/file.hpp>
+#include <leatherman/execution/execution.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+
+using namespace std;
+using namespace boost::filesystem;
+namespace bs = boost::system;
+namespace lth_file = leatherman::file_util;
+using namespace leatherman::util;
+
+namespace whereami { namespace sources {
+
+    std::string dmi_base::bios_vendor() const
+    {
+        return data_ ? data_->bios_vendor : "";
+    }
+
+    std::string dmi_base::board_manufacturer() const
+    {
+        return data_ ? data_->board_manufacturer : "";
+    }
+
+    std::string dmi_base::board_product_name() const
+    {
+        return data_ ? data_->board_product_name : "";
+    }
+
+    std::string dmi_base::product_name() const
+    {
+        return data_ ? data_->product_name : "";
+    }
+
+    std::string dmi_base::manufacturer() const
+    {
+        return data_ ? data_->manufacturer : "";
+    }
+
+    dmi::dmi()
+    {
+        collect_data();
+    }
+
+    void dmi::collect_data()
+    {
+        if (!collect_sys_data()) {
+            collect_dmidecode_data();
+        }
+    }
+
+    string dmi::sys_path(string const& filename) const
+    {
+        return "/sys/class/dmi/id/" + filename;
+    }
+
+    bool dmi::collect_sys_data()
+    {
+        // Check that /sys/class/dmi exists
+        bs::error_code ec;
+        if (is_directory(sys_path(), ec)) {
+            data_.reset(new dmi_data);
+            data_->bios_vendor = read_file(sys_path("bios_vendor"));
```

Comment:
This is going to propagate an exception out if we for some reason can't read these files (non-root?)

I think that's OK for now - we'll want to figure out how to report source failures to detectors somehow, so that they can properly raise warnings when results might be inaccurate due to missing/broken system data.